### PR TITLE
Port "Verify symbol type when reporting HAA0602"

### PR DESCRIPTION
**Pull request by genlu**
Ported from https://github.com/microsoft/RoslynClrHeapAllocationAnalyzer/pull/79. 
Originally implemented by @cezarypiatek

FYI @mjsabby

### CODE REVIEW

File: src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs [hunk 1]
```diff
@@ -325,12 +325,22 @@ private static void CheckDelegateCreation(SyntaxNode node, TypeInfo typeInfo, Se
...
+                if (IsStructInstanceMethod(node!, semanticModel, cancellationToken))
                 {
                     reportDiagnostic(Diagnostic.Create(DelegateOnStructInstanceRule, location, EmptyMessageArgs));
                 }
             }
         }
+
+        private static bool IsStructInstanceMethod(SyntaxNode? node, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            if (node.IsKind(SyntaxKind.AnonymousMethodExpression) || node.IsKind(SyntaxKind.ParenthesizedLambdaExpression))
```
Comment (sharwell) [inline src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs]: ❓ Is this still correct if the lambda captures `this`?

File: src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs [hunk 2]
```diff
@@ -325,12 +325,22 @@ private static void CheckDelegateCreation(SyntaxNode node, TypeInfo typeInfo, Se
...
+            {
+                return false;
+            }
+
+            var symbolInfo = semanticModel.GetSymbolInfo(node, cancellationToken).Symbol;
+            return symbolInfo != null && symbolInfo.Kind == SymbolKind.Method && symbolInfo.IsStatic == false && symbolInfo.ContainingType?.IsValueType == true;
```
Comment (sharwell) [inline src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs]:
```suggestion
            return symbolInfo?.Kind == SymbolKind.Method && symbolInfo.IsStatic == false && symbolInfo.ContainingType?.IsValueType == true;
```

File: src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs (hunk 1 above)
Comment (genlu) [reply src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs]: That would be a boxing by itself, right? I think this check is only for wrapping an instance method on struct

Comment (sharwell) [inline src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs]: 💭 Is this missing `SyntaxKind.SimpleLambdaExpression`?

File: src/PerformanceSensitiveAnalyzers/UnitTests/TypeConversionAllocationAnalyzerTests.cs [hunk 3]
```diff
@@ -686,5 +692,120 @@ public void SomeMethod()
...
+public struct MyStruct
+{
+    [PerformanceSensitive(""uri"")]
+    public void Testing()
+    {
+        var ints = new[] { 5, 4, 3, 2, 1 };
+        Array.Sort(ints, delegate(int x, int y) { return x - y; });
+        Array.Sort(ints, (x, y) => x - y);
+        DoSomething(() => throw new Exception());
+        DoSomething(delegate() { throw new Exception(); });
```
Comment (sharwell) [inline src/PerformanceSensitiveAnalyzers/UnitTests/TypeConversionAllocationAnalyzerTests.cs]: 💡 This needs a test with the simple lambda syntax as well (`x => throw new Exception()`)

File: src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs (hunk 1 above)
Comment (sharwell) [reply src/PerformanceSensitiveAnalyzers/CSharp/TypeConversionAllocationAnalyzer.cs]: It turns out it's not possible to box from within a lambda in a struct, so we're good there.

**Pull request merged**